src/contracts/core/Slasher.sol

```diff
@@ -1,102 +1,368 @@
+        require(
+            bipsToReduce < 2 ** 31, "Slasher.reduceRequestedBipsToSlash: bipsToReduce must be less than minimum int32"
+        );
+        _modifyRequestedBipsToSlash(operator, operatorSetID, strategies, epoch, -int32(bipsToReduce));
+    }
 
-    function freezeOperator(address) external {}
+    /**
+     * @notice Permissionlessly called to execute slashing of a given list of
+     * strategies for a given operator, for the latest unslashed epoch
+     *
+     * @param operator the operator to slash
+     * @param strategies the list of strategies to execute slashing for
+     * @param epoch the epoch in which the slashing requests to execute were made
+     */
+    function executeSlashing(address operator, IStrategy[] memory strategies, uint32 epoch) external {
+        require(
+            EpochUtils.currentEpoch() > EpochUtils.getMinExecutionEpochFromRequestEpoch(epoch),
+            "Slasher.executeSlashing: current epoch must be greater than the minimum execution epoch"
+        );
+        for (uint256 i = 0; i < strategies.length; ++i) {
+            IStrategy strategy = strategies[i];
+            // make sure the slashing request is in order and overwrite the last executed slashing request ID
+            SlashingRequest memory slashingRequest = slashingRequests[operator][strategy][epoch];
+            require(
+                slashingRequestIds[operator][strategy].lastExecutedSlashingRequestId + 1 == slashingRequest.id,
+                "Slasher.executeSlashing: must execute slashings in order"
+            );
+            slashingRequestIds[operator][strategy].lastExecutedSlashingRequestId = slashingRequest.id;
 
-    function resetFrozenStatus(address[] calldata) external {}
+            // truncate to BIPS_FACTOR_SQUARED if it exceeds it
+            if (slashingRequest.slashingRate > SlashingAccountingUtils.BIPS_FACTOR_SQUARED) {
+                slashingRequest.slashingRate = SlashingAccountingUtils.BIPS_FACTOR_SQUARED;
+            } else if (slashingRequest.slashingRate == 0) {
+                // if the slashing rate is 0 due to cancellation, skip the slashing
+                continue;
+            }
+            
+            uint64 scalingFactorBefore = shareScalingFactor(operator, strategy);
+            uint64 scalingFactorAfter = SlashingAccountingUtils.findNewScalingFactor({
+                scalingFactorBefore: scalingFactorBefore,
+                rateToSlash: slashingRequest.slashingRate
+            });
+            // update storage to reflect the slashing
+            slashedEpochHistory[operator][strategy].push(epoch);
+            _shareScalingFactor[operator][strategy] = scalingFactorAfter;
+            // TODO DONE?: note that this behavior risks off-by-one errors. it needs to be clearly defined precisely how the historical storage is supposed to work
+            slashingRequest.scalingFactor = scalingFactorAfter;
```

Comment:
This doesn't update anything, storage-wise.

But - do we even want to do an update here? It seems like we're done with this `slashingRequest`